test/model/message_test.dart

```diff
@@ -123,6 +127,179 @@ void main() {
     });
   });
 
+  group('edit-message methods', () {
+    late PerAccountStore store;
+    late FakeApiConnection connection;
+    late StreamMessage message;
+
+    Future<void> prepare() async {
+      store = eg.store();
+      connection = store.connection as FakeApiConnection;
+      message = eg.streamMessage();
+      await store.addMessage(message);
+    }
+
+    test('smoke', () => awaitFakeAsync((async) async {
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(
+        json: UpdateMessageResult().toJson(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      check(connection.takeRequests()).single.isA<http.Request>()
+        ..method.equals('PATCH')
+        ..url.path.equals('/api/v1/messages/${message.id}')
+        ..bodyFields.deepEquals({
+          'content': 'new content',
+        });
+
+      async.elapse(Duration(milliseconds: 500));
+      // Mid-request
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isFalse();
+
+      async.elapse(Duration(milliseconds: 500));
+      // Request has succeeded; event hasn't arrived
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isFalse();
+
+      await store.handleEvent(eg.updateMessageEditEvent(message));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+    }));
+
+    test('request fails', () => awaitFakeAsync((async) async {
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+    }));
+
+    test('request fails; take failed edit', () => awaitFakeAsync((async) async {
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+
+      check(store.takeFailedMessageEdit(message.id)).equals('new content');
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+    }));
+
+    test('takeFailedMessageEdit throws StateError when nothing to take', () => awaitFakeAsync((async) async {
```

Comment:
I think a similarly useful test would be repeated `editMessage` calls that operate on the same message.